fv3core/utils/future_stencil.py

```diff
@@ -0,0 +1,307 @@
+
+    def _set_buffer(self, buffer: np.ndarray):
+        self._window.Lock(rank=0)
+        self._window.Put(buffer, target_rank=0, target=self._get_target())
+        self._window.Unlock(rank=0)
+
+    def _get_buffer(self, node_id: int = -1) -> np.ndarray:
+        buffer = np.empty(self._buffer_size, dtype=self._np_type)
+        self._window.Lock(rank=0)
+        self._window.Get(buffer, target_rank=0, target=self._get_target(node_id))
+        self._window.Unlock(rank=0)
+
+        return buffer
+
+
+def future_stencil(
+    backend: Optional[str] = None,
+    definition: Optional[Callable] = None,
+    *,
+    externals: Optional[Dict[str, Any]] = None,
+    wrapper: Optional[Callable] = None,
+    rebuild: bool = False,
+    **kwargs: Any,
+):
+    """
+    Create a future stencil object with deferred building in a distributed context
+
+    Parameters
+    ----------
+        backend : `str`
+            Name of the implementation backend.
+
+        definition : `None` when used as a decorator, otherwise a `function` or a
+                     `:class:`gt4py.StencilObject`
+            Function object defining the stencil.
+
+        externals: `dict`, optional
+            Specify values for otherwise unbound symbols.
+
+        rebuild : `bool`, optional
+            Force rebuild of the :class:`gt4py.StencilObject` even if it is
+            found in the cache. (`False` by default).
+
+        **kwargs: `dict`, optional
+            Extra backend-specific options. Check the specific backend
+            documentation for further information.
+
+    Returns
+    -------
+        :class:`FutureStencil`
+            Wrapper around an instance of the dynamically-generated subclass
+            of :class:`gt4py.StencilObject`.
+    """
+
+    def _decorator(func):
+        device_sync = kwargs.pop("device_sync", None)
+        backend_opts = {"device_sync": device_sync} if device_sync else {}
+        builder = (
+            StencilBuilder(func)
+            .with_backend(backend)
+            .with_externals(externals or {})
+            .with_options(
+                name=func.__name__,
+                module=func.__module__,
+                rebuild=rebuild,
+                backend_opts=backend_opts,
+                **kwargs,
+            )
+        )
+        return FutureStencil(builder, wrapper)
+
+    if definition is None:
+        return _decorator
+    return _decorator(definition)
+
+
+class FutureStencil:
+    """
+    A stencil object that is compiled by this node or another in a distributed context.
+    """
+
+    _id_table = StencilTable()
+
+    def __init__(
+        self,
+        builder: Optional["StencilBuilder"] = None,
+        wrapper: Optional[Callable] = None,
+    ):
+        self._builder: Optional["StencilBuilder"] = builder
+        self._stencil_object: Optional[StencilObject] = None
+        self._sleep_time: float = 0.05
+        self._timeout: float = 180.0
```

Comment:
We have stencils that take more than 180s, e.g. physics. I have no good idea apart from raising that limit though :/

Reply:
Technically we would need a ping to the node compiling to make sure it's alive

Reply:
That could potentially be done, but for now I would be in favor of increasing the timeout.

Reply:
Can we comment that in though? So at least it's commented in-code

Reply:
Absolutely.